[JENKINS-60734] Replace jsbuilder with webpack

fqueiruga opened the pull request:
See [JENKINS-60734](https://issues.jenkins-ci.org/browse/JENKINS-60734).

This PR updates the frontend toolchain to use webpack. It is supposed to be a non-breaking change that shouldn't affect plugins. The changes are limited to different static assets being built **only on Jenkins core**.

### Proposed changelog entries

* Major improvement: frontend toolchain upgraded to use webpack. Enabled babel to transpile certain JS modules.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

* @halkeye
* @uhafner 
* @timja 

### TODOs:

- [x] Add documentation of how to build on development mode (`yarn start`).
- [x] Migrate and enable the JS unit tests
- [x] Migrate and enable the linter (maybe replace jshint with eslint)

### Important changes

- Stops using libraries from the js-libs repo. Bring jquery, handlebars, etc from NPM instead.
    - One exception is the bootstrap-detached library. I needed to inline it because I wasn't able to load the actual bootstrap JS code.
- Font files were moved to the `war/src/main/fonts` folder, so that they can be processed by the LESS loader. They are copied to their previous location at build time, so this is not a breaking change.

Screenshot of build results:
![image](https://user-images.githubusercontent.com/5738588/72170325-00a1b300-33d1-11ea-967c-90a5caad6f15.png)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

fqueiruga committed:
set up webpack

fqueiruga committed:
add pluginsetupwizard.less to webpack

fqueiruga committed:
Migrates all JS code to webpack

- moves the fonts back to the war/src/main/webapp/css folder

fqueiruga committed:
Uses the fonts just once. They are copied to the webapp/css folder on build time

fqueiruga committed:
Separates vendor scripts

fqueiruga committed:
Adds source build to maven and adds a yarn start command

fqueiruga committed:
Remove vscode config

fqueiruga committed:
Remove debug console statement

fqueiruga committed:
Bumping the Jenkins version to make it work on CI

fqueiruga committed:
Removes gulp, js-builder and other unused packages

fqueiruga committed:
Spacing and comment fixes

fqueiruga committed:
Revert more whitespace-only changes

fqueiruga committed:
Adds some codestyle changes

fqueiruga committed:
Removes more whitespace changes

fqueiruga committed:
Undo changes to the setup wizard test file

fqueiruga committed:
Fix: broken reference on the setup wizard progress page

fqueiruga commented on war/package.json:
```diff
@@ -9,18 +9,38 @@
   },
   "private": true,
+  "scripts": {
+    "copy-fonts": "cp -rp src/main/fonts/* src/main/webapp/css",
```
Question: is Jenkins built on an environment where this `cp` command would fail?

halkeye replied:
Yea. It's tested on windows.

Theres a webpack copy plugin that works really well

fqueiruga replied:
I have added the webpack copy plugin. Thanks!

res0nance commented:
Maybe it should be highlighted that this does not change anything for the user despite being a major-rfe.

fqueiruga commented:
> Maybe it should be highlighted that this does not change anything for the user despite being a major-rfe.

How could we do it, other than updating the changelog entry?

res0nance commented:
> > Maybe it should be highlighted that this does not change anything for the user despite being a major-rfe.
> 
> How could we do it, other than updating the changelog entry?

Probably just make sure the changelog entry mentions it.

I'm sure this will light the way for easier UI changes but I think we should ensure that users do not get confused.

daniel-beck commented:
This is what the "Internal" prefix is for: https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/_STYLEGUIDE.adoc#labels

If I understand the labels to this issue correctly, that will be done automatically already.

daniel-beck commented on war/src/main/js/handlebars-helpers/id.js:
```diff
@@ -0,0 +1,3 @@
+export default function id(str) {
+	return (''+str).replace(/\W+/g, '_');
```
Does it matter that potentially multiple values share the same ID with this replacement scheme?

fqueiruga replied:
TBH I don't know. I limited the scope of this PR to just maintain current behaviour. I extracted the function as it is.

fqueiruga replied:
I'm resolving this one, feel free to reopen it if you disagree!

fqueiruga committed:
Moved the font copying to webpack

- It was previously done using cli, which could fail on windows
- Added plugin to clean up webpack build folder
- Cleaup on the package.json

halkeye commented on war/.jshintrc.json:
```diff
@@ -0,0 +1,11 @@
+{
```
While we are piling on wishlist items, can you switch to eslint because... well i have actually no solid reasons, i just like it a lot better and it has way more configuration options. https://github.com/jenkins-infra/plugin-site/blob/master/.eslintrc.js as an example

fqueiruga replied:
I'd also suggest going for eslint. My concern would be that there would suddenly be 100s of new lint warnings. From what I see I would also need to disable all styling rules. 

Do you think there would be any problem in this regard?

halkeye replied:
Just set everything to warn to start. When fix the warnings in a future pr

fqueiruga replied:
@halkeye is there any eslint preset you recommend that can be good fit this project. I have only used the react-app one for the past year or so (which I think it's quite good).

halkeye replied:
I don't have strong opinions about it. But if you run eslint --init you can infer all the rules based on the current codebase, so would allow things to be linted without making changes

fqueiruga replied:
I have enabled eslint and stylelint with default configs

halkeye commented on war/.babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["@babel/preset-env"],
```
if your going to use env (which is +1 from me) make sure you have browserlist defined, i'm pretty sure core needs to support at least ie11

fqueiruga replied:
Will do! Do you think I should make the move to env more visible? I was thinking that it could deserve it's own changelog entry.

halkeye replied:
I kinda want to say delay using babel for a future pr, 

Scratch that.

https://jenkins.io/doc/administration/requirements/web-browsers/ needs to be followed. After that your probably fine. So if your browser list matches up then probably no real concern.

It could potentially break JavaScript for older browsers, but since we don't say we support them, and people using ie10 are unlikely to upgrade Jenkins, I think your fine, but a note certainly wouldn't hurt

Cc @oleg-negashev who updated the list last

oleg-nenashev replied:
+1 for following the policy or noting a breaking change if needed

fqueiruga replied:
I'll write an explanation about the change to babel. Is there any place other than the JIRA ticket and changelog entry where I would need to write about it (or where can I consult the policy).

FWIW, I have tested on a IE VM and it seems to work on IE10, but I agree a disclaimer can be good just in case.

timja replied:
possibly here:
https://jenkins.io/doc/developer/building/
or somewhere near here:
https://jenkins.io/doc/developer/views/

fqueiruga replied:
Some plugins include frontend dev instructions on the package README (https://github.com/jenkinsci/pipeline-stage-view-plugin/tree/master/ui). 

I guess this is not the way that's done on Jenkins core. Are doc pages on jenkins.io preferred?
If this is this the case, I suppose I'd need to create a PR on https://github.com/jenkins-infra/jenkins.io right?

timja replied:
jenkins.io would be preferred for docs that affect multiple plugins, developer experience, and guides on how to do things.

if it's very specific to core, it could just be documented in this repo

and yes a PR to that repo if going to jenkins.io

fqueiruga replied:
I have updated the CONTRIBUTING.md with instructions specific to frontend work: https://github.com/jenkinsci/jenkins/pull/4425/commits/b58ea61223a16868a263ce06f0c76dbc2d7686c0

fqueiruga replied:
Resolving this one, I'll send a PR to the jenkins.io repo if more docs are needed.

fqueiruga committed:
Adds browserslist config (defaults + ie11)

batmat commented:
I just marked this PR as work-in-progress.

It is deemed close to ready, hence why it was filed for early feedback, and the provided feedback is being actively addressed.

Thanks Félix!

varyvol commented on war/.jshintrc.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "curly": true,
```
I guess it can probably be found easily in the documentation, but maybe a comment explaining what each option means would be helpful. This is a quite big project and not everybody could be knowledgeable about this.

timja replied:
this is a json file, is it possible to add comments?

fqueiruga replied:
You cannot leave comments in these files.

This config file belongs to jshint, it's the default one. I'll remove it as I'll try to replace jshint with eslint. On eslint, I'm partial too looking for a standard config, so that developers don't need to worry about the reasoning for each specific rule.

varyvol commented on war/src/main/js/handlebars-helpers/ifneq.js:
```diff
@@ -0,0 +1,5 @@
+export default function ifneq(o1, o2, options) {
```
Why are these functions necessary?

varyvol replied:
I mean: they were in a another file before. But do you think there could be places where they are used from outside?

fqueiruga replied:
Handlebars templates are compiled and processed by webpack. In order to do this, Handlebars needs to know which helpers are available at compile time. That's why I extracted as many helpers as possible to this dir. Note that they can still be 

Every helper that is registered at runtime must be listed on the _webpack.config.js file_. Only a small set of helpers are still registered this way, and the reason for that is that they break when extracted (due to their use of closure scope).

fqueiruga replied:
Resolving this one. Feel free to unresolve it if you think further discussion is needed.

varyvol commented on war/src/test/js/widgets/config/mocks.js:
```diff
@@ -1,17 +1,18 @@
+// TODO: remove or do something about this when reenabling the unit tests
```
Is this comment still valid?

fqueiruga replied:
Not anymore :D good catch, I'll remove it

varyvol commented on war/yarn.lock:
```diff
@@ -2,31 +2,904 @@
 # yarn lockfile v1
 
 
+"@babel/cli@^7.7.4":
```
Is this file necessary? It seems like something automatically generated?

fqueiruga replied:
This file is autogenerated, so I guess _yarn_ considers it as a needed dependency.

timja replied:
It’s a lock file for repeatable builds and should be checked in

varyvol replied:
I'm not sure of understanding. It's autogenerated but it needs to be checked in?

fqueiruga committed:
Migrates unit tests to Jest

- Jest was chosen because it's the most feature-complete one
- Fixes tests so that they would work with the new JS modules
- Removed the gulpfile unused test dependencies
- Enabled tests in maven

fqueiruga committed:
Adds a reporter because jest will output messages to stderr

fqueiruga commented:
I have migrated and enabled the unit tests. I have chosen [jest](https://jestjs.io/) as the runner for several reasons:

- It has the widest ammount of features (does not need an assertion library and allows mocking modules).
- It's compatible with the Jasmine syntax of the current tests.

I needed to do lots of refactoring on the tests. Their behaviour was coupled to the old module systems.

fqueiruga committed:
Remove outdated comment

batmat commented:
PR is failing with these test failures:

![image](https://user-images.githubusercontent.com/223853/72681832-a3d88380-3ac7-11ea-93fa-41d22e693bfb.png)


BUT AFAICT these failures should be unrelated, since I can see the very same on `master` on an older commit
![image](https://user-images.githubusercontent.com/223853/72681823-8c00ff80-3ac7-11ea-9248-a16c0c7c27af.png)

batmat commented:
AFAIU, https://github.com/jenkinsci/jenkins/pull/4435 was merged 4 hours to disable these tests, so we should expect now a green build (or no failure anymore from these tests at least).

timja commented:
> "its" current. (I'm not sure why GitHub is not allowing me to make suggestions.

because you're commenting on a commit.

uhafner commented on core/src/main/resources/lib/layout/html.jelly:
```diff
@@ -161,6 +161,8 @@ THE SOFTWARE.
       <script src="${resURL}/scripts/msie.js" type="text/javascript"/>
     </j:if>
 
+    <!-- Load vendor scripts common to all JS modules -->
```
Maybe make clear that this file is generated during the build.

fqueiruga replied:
I have updated the comment on https://github.com/jenkinsci/jenkins/pull/4425/commits/99832f1b9460d4c60cd4ca84762adebba0aacdc6 .

Marking it as resolved. Feel free to unresolve it if you'd like further change!

uhafner commented on war/src/main/js/add-item.js:
```diff
@@ -1,5 +1,4 @@
-// Initialize all modules by requiring them. Also makes sure they get bundled (see gulpfile.js).
-var $ = require('jquery-detached').getJQuery();
```
Do plugins that depend on `jquery-detached` need updates as well?

fqueiruga replied:
They don't. The code still works for plugins that use js-builder, js-modules and jquery-detached. With this update, Jenkins core just bundles its own dependencies on the vendors.js folder. Plugins using the older system still work as expected. 

FWIW, I have manually verified that the pipeline stageview and warnings-ng still work.

uhafner replied:
Yes, I also checked that my plugin is still working 🎉.

fqueiruga replied:
Closing this one then!

uhafner reviewed (commented):
Does this PR effect Jenkins Plugins that use jQuery in the old way (or use jQuery via https://github.com/jenkinsci/jquery-plugin/ or https://github.com/jenkinsci/jquery3-api-plugin)?

fqueiruga commented:
> Does this PR effect Jenkins Plugins that use jQuery in the old way (or use jQuery via https://github.com/jenkinsci/jquery-plugin/ or https://github.com/jenkinsci/jquery3-api-plugin)?

It shouldn't. The behaviour does not fundamentally change. It still generates a JS bundle that does not affect other JS code outside of it. I'll do some manual tests to verify that these plugins work.

fqueiruga committed:
Enables linting for JS and LESS files

- Uses eslint for JS and stylelint for LESS
- Uses default configs
- Default severity set to warn to avoid breaking current builds

fqueiruga committed:
Fixes typo

fqueiruga committed:
Adds clarification that the vendors.js file is built by webpack

fqueiruga committed:
Add development instructions and bump the node versions

fqueiruga commented:
I have added instructions to work on the frontend assets on the CONTRIBUTING.md file. They were added on this commit: https://github.com/jenkinsci/jenkins/pull/4425/commits/b58ea61223a16868a263ce06f0c76dbc2d7686c0

fqueiruga commented:
I think the main work has all been completed. I'd like to request that this PR is moved out of WIP status. Thanks!

timja reviewed (approved):
Please rename the title to remove wip

alecharp commented on CONTRIBUTING.md:
````diff
@@ -49,6 +51,21 @@ mvn -pl war jetty:run
+cd war; yarn start
+```
+
+
````
```suggestion
```

alecharp reviewed (approved):
Thanks @fqueiruga for this !

halkeye commented:
```
$ ag js-builder
pom.xml
580:        See https://github.com/tfennelly/jenkins-js-builder#maven-integration
```

want to clean up that pom.xml?

uhafner reviewed (approved):
I'm almost sure that I already sent a review...

halkeye commented on war/src/main/js/api/securityConfig.js:
```diff
@@ -1,13 +1,12 @@
  */
-
-var jenkins = require('../util/jenkins');
+import jenkins from '../util/jenkins';
```
for consistency can you move the two requires up to import at the top?

fqueiruga replied:
I have done it in this commit: https://github.com/jenkinsci/jenkins/pull/4425/commits/6cccb699697374b54470db95c2f6043168b88c8d

What do you think of the result?

halkeye commented:
So I started to try and make a more explicit eslint file that matches all the current implementation, but I'm noticing a lot of inconistencies

So you switched everything to use imports instead of require. Any reason why you did that? There's a bunch of requires left behind. and using `exports` directly isn't supported (though does sorta work).

I would prefer everything switch to imports, or everything amd, not introducing a mix of both.

halkeye commented:
> /home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/main/js/util/jenkinsLocalStorage.js
>   7:1  error  'exports' is not defined  no-undef
>  14:1  error  'exports' is not defined  no-undef
>  21:1  error  'exports' is not defined  no-undef
>  29:1  error  'exports' is not defined  no-undef

should be export (As in import/export) or module.exports (as in require)

halkeye commented:
https://gist.github.com/2c02d54345a7219df3faecbf7e1b6e5b
^^ This is eslint init infer of the content of src/main/**/*.js

The current file uses a plugin to disable all errors, so its kinda an all or nothing.

with this file it does catch a few tiny things
```
/home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/main/js/add-item.js
  38:35  error  Unnecessary escape character: \&  no-useless-escape
  38:57  error  Unnecessary escape character: \&  no-useless-escape

/home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/main/js/api/securityConfig.js
  22:2  error  Unnecessary semicolon  no-extra-semi
  37:2  error  Unnecessary semicolon  no-extra-semi
  50:2  error  Unnecessary semicolon  no-extra-semi

/home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/main/js/pluginSetupWizardGui.js
   532:9   error  'restartRequired' is assigned a value but never used  no-unused-vars
  1220:13  error  'txt' is assigned a value but never used              no-unused-vars

/home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/main/js/util/jenkins.js
   25:7   error  'Hash' is not defined                                     no-undef
   31:11  error  'Hash' is not defined                                     no-undef
   44:4   error  'Hash' is not defined                                     no-undef
  148:14  error  'Proxy' is already defined as a built-in global variable  no-redeclare

/home/halkeye/go/src/github.com/jenkinsci/jenkins/war/src/test/js/pluginSetupWizard.spec.js
  233:9  warning  Test has no assertions                                        jest/expect-expect
  306:5  warning  Test has no assertions                                        jest/expect-expect
  320:9  warning  Test has no assertions                                        jest/expect-expect
  446:9  warning  Test has no assertions                                        jest/expect-expect
  452:8  error    Test title is used multiple times in the same describe block  jest/no-identical-title
```

fqueiruga commented on war/pom.xml:
```diff
@@ -41,8 +41,8 @@ THE SOFTWARE.
     <port>8080</port><!-- HTTP listener port -->
-    <node.version>6.10.2</node.version>
-    <yarn.version>0.24.5</yarn.version>
+    <node.version>12.14.0</node.version>
```
Done

fqueiruga commented on core/src/main/resources/lib/layout/layout.jelly:
```diff
@@ -173,6 +173,8 @@ THE SOFTWARE.
       <script src="${resURL}/scripts/msie.js" type="text/javascript"/>
     </j:if>
 
+    <!-- Load vendor scripts common to all JS modules -->
```
I have added this clarification as per Ulli's feedback https://github.com/jenkinsci/jenkins/pull/4425/commits/99832f1b9460d4c60cd4ca84762adebba0aacdc6

fqueiruga commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -1266,5 +1266,7 @@ var createPluginSetupWizard = function(appendTarget) {
+var exports = {};
 // export wizard creation method
 exports.init = createPluginSetupWizard;
+export default exports;
```
I guess I can change it to 
```js
export default { init: createPluginSetupWizard };
```

alecharp replied:
yes correct. But it's good as it is. We will in the future I guess revisit some of the JS / UI code.

fqueiruga replied:
Done!

fqueiruga committed:
Update CONTRIBUTING.md

Co-Authored-By: Adrien Lecharpentier <adrien.lecharpentier@gmail.com>

fqueiruga committed:
Migrates from requires to imports

- Fixes some related linter errors in the way

timja commented on war/src/test/js/widgets/config/tabbar.spec.js:
```diff
@@ -1,31 +1,79 @@
+        }, htmlConfigTabbedContent);
     });
 
+    // This test may be deterministic, as it breaks if jest.resetModules is called on afterEach
```
did you mean not deterministic?
```suggestion
    // This test may not be deterministic, as it breaks if jest.resetModules is called on afterEach
```

fqueiruga replied:
Yeah, exactly

fqueiruga committed:
Move module-level comments to underneath the imports

fqueiruga committed:
Update war/src/test/js/widgets/config/tabbar.spec.js

Co-Authored-By: Tim Jacomb <t.jacomb@kainos.com>

alecharp commented:
@halkeye I'm not sure we should address `eslint` warnings here if they don't break the build. 

I would prefer to have this PR about having the tool ready to be used, and then we can discuss about which rules are more important and from there create PR where we activate one rule at a time and fix the warnings / errors the rule brings up. 

The PR is already very large as it is.

fqueiruga commented:
@halkeye I have implemented most of your suggestions. I went all-in using `import`, no more requires laying around except where needed (in tests, basically). No trace of `module.exports` either.

I did not yet apply the changes to the eslint config as you listed on the gist. My goal on this PR was to have a minimal setup working. While the eslint config can definitely be improved I would suggest to move it to another PR.

fqueiruga committed:
Adds .eslintignore file and changes the way eslint is run

fqueiruga committed:
Fix: broken interface of the jenkins object yielded by the setupWizard

- Callbacks registered with `onSetupWizardInitialized` are yielded a `wizard` parameter, that contained a reference to the jenkins object.
- Some plugins called the method `jenkins.initHandlebars` from this callback, to receive a Handlebars intance.
- Added a unit test to ensure this regression does not happen again

timja commented:
6 approvals, and no existing comments to be addressed in this PR that I can see, adding ready for merge

batmat commented:
Thanks Tim!

Yes, agreed. I just manually copied the [new part added by Oleg from the PR template for maintainers](https://raw.githubusercontent.com/jenkinsci/jenkins/master/.github/PULL_REQUEST_TEMPLATE.md) and made a pass to check each item.
Note that I did not check the last item about backporting to LTS on purpose: clearly, this change is not eligible to a backport given its size :-).

Given everything is OK, and as Tim said we have much more than the minimum number of approvals, I am merging this PR.

Goes without saying, we will handle fallouts as need be.

:shipit:

halkeye commented:
>  Note that I did not check the last item about backporting to LTS on purpose: clearly, this change is not eligible to a backport given its size :-).

also it gives no benefit to LTS, its purely a developer change (hopefully)

basil commented on war/.eslintignore:
```diff
@@ -0,0 +1,20 @@
+src/main/webapp/css/google-fonts
+src/main/webapp/css/icomoon
+
+# External scripts
+src/main/webapp/scripts/yui
+src/main/webapp/scripts/prototype.js
+src/main/webapp/scripts/*.js
```
But `src/main/webapp/scripts/hudson-behavior.js` is an internal script, not an external script.

https://github.com/jenkinsci/jenkins/pull/8012
